### PR TITLE
Add dashboard report

### PR DESCRIPTION
**Pull request by 2e3s**

### CODE REVIEW

File: spec/paths/reports@dashboard.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+get:
```
Comment (slavcodev) [inline spec/paths/reports@dashboard.yaml]: GET?

File: spec/paths/reports@dashboard.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      format: date-time
+    - name: periodEnd
+      in: query
+      description: Report end day
+      required: true
+      type: string
+      format: date-time
+    - name: body
+      in: body
+      description: Subscription resource
```
Comment (slavcodev) [inline spec/paths/reports@dashboard.yaml]: seems, is not relevant description

File: spec/paths/reports@dashboard.yaml (hunk 1 above)
Comment (2e3s) [reply spec/paths/reports@dashboard.yaml]: You're right, copypaste error.

**Comment by 2e3s** [comment]
@slavcodev why `GET`? It would be less convenient here it is a more complex API (unusable manually). For instance, Elasticsearch and many others also use `POST` for complex queries.

The dashboard report just combines Transaction Report and Transaction Chart. 1 endpoint per metric would be highly supoptimal because now the data is fetched for all the metrics at once and then divided per specified metrics.

I thought of another idea after I finished this: using Transaction Chart endpoint and Transactions Report to fetch data from the frontend directly for all metrics at once. It would involve a few more API calls and more frontend work. I'm not sure what is better. When I did this one it seemed to me possibly better for the future needs and more extendable (this API doesn't affect others).

**Comment by slavcodev** [comment]
> It would involve a few more API calls and more frontend work (a bit less performant)

I"m not sure, but one big request may be less performant, than many small.

**Comment by slavcodev** [comment]
> why GET? It would be less convenient here it is a more complex API (unusable manually). For instance, Elasticsearch and many others also use POST for complex queries. So there should be rationale to overcome the inconvenience.

This is just my knowledges. POST is for commands, and is not for queries.
There is should be rationale to use unappropriated method too.

**Comment by 2e3s** [comment]
> I"m not sure, but one big request may be less performant, than many small.

I see how additional requests can decrease speed, how can they increate it?

**Comment by slavcodev** [comment]
HTTP requests are data transport, right? So big amount of data is sent longer than small. In the same time many requests are async sending in parallel.

**Comment by 2e3s** [comment]
Running it in parallel may not compensate the overhead on new requests (it's network + connections + bootstrapping our backend with no reason) and I believe it won't. Also it's less convenient to see the reports usage (dashboard would be seen. Also it requires more work from F-E for a few async requests.
I think it may be very **DRY** but in fact much worse by all means.

> POST is for commands, and is not for queries.

Report can be named "a command" too. There is a restriction for GET against non-idempotent requests, but there is no restrictions on POST.

So we have 2 alternatives:
`POST experimental/reports/dashboard?periodStart=2016-10-01T00%3A00%3A00Z&periodEnd=2016-11-24T00%3A00%3A00Z

{
  "metrics": {
    "approval-rate": {
      "type": "revenue",
      "period": "day",
      "segments": {
        "facebook": {
          "criteria": {
            "op": "equals",
            "path": "/bin",
            "value": "123465"
          }
        }
      }
    }
  },
	"sales": {
	  "type": "salesAllCount",
	  "period": "day",
	  "segments": {
	    "facebook": {
	      "criteria": {
	        "op": "equals",
	        "path": "/bin",
	        "value": "123465"
	      }
	    }
	  }
	}
}
`
or
`GET experimental/reports/dashboard?periodStart=2016-10-01T00%3A00%3A00Z&periodEnd=2016-11-24T00%3A00%3A00Z&body=%7B%0D%0A++%22metrics%22%3A+%7B%0D%0A++++%22approval-rate%22%3A+%7B%0D%0A++++++%22type%22%3A+%22revenue%22%2C%0D%0A++++++%22period%22%3A+%22day%22%2C%0D%0A++++++%22segments%22%3A+%7B%0D%0A++++++++%22facebook%22%3A+%7B%0D%0A++++++++++%22criteria%22%3A+%7B%0D%0A++++++++++++%22op%22%3A+%22equals%22%2C%0D%0A++++++++++++%22path%22%3A+%22%2Fbin%22%2C%0D%0A++++++++++++%22value%22%3A+%22123465%22%0D%0A++++++++++%7D%0D%0A++++++++%7D%0D%0A++++++%7D%0D%0A++++%7D%0D%0A++%7D%2C%0D%0A%09%22sales%22%3A+%7B%0D%0A%09++%22type%22%3A+%22salesAllCount%22%2C%0D%0A%09++%22period%22%3A+%22day%22%2C%0D%0A%09++%22segments%22%3A+%7B%0D%0A%09++++%22facebook%22%3A+%7B%0D%0A%09++++++%22criteria%22%3A+%7B%0D%0A%09++++++++%22op%22%3A+%22equals%22%2C%0D%0A%09++++++++%22path%22%3A+%22%2Fbin%22%2C%0D%0A%09++++++++%22value%22%3A+%22123465%22%0D%0A%09++++++%7D%0D%0A%09++++%7D%0D%0A%09++%7D%0D%0A%09%7D%0D%0A%7D%0D%0A`

I feel like there is a flaw in such GET

**Comment by slavcodev** [comment]
I do not feel like you. What flaw do you see there?

**Comment by 2e3s** [comment]
It looks misplaced, unreadable, undocumentable, inconvenient to work with, unlike the json body being posted with POST. It's something subjective though. I can't find any restriction for POST to work with searches/reports. What document should I rely on when changing POST to GET here?

**Comment by 2e3s** [comment]
I don't like it (by the aforementioned reasons) but I have decided to remake it on `GET`.

**Comment by slavcodev** [comment]
> I don't like it (by the aforementioned reasons) but I have decided to remake it on GET.

What make you decide this, contrary to your feelings? I shared my opinion, my **one** opinion. Seems nobody else does support it. It's your task, you who know better the requirement.

**Comment by slavcodev** [comment]
> It looks misplaced, unreadable, undocumentable, inconvenient to work with

Are you going to read this requests or application? Seems it is simple to document this query param in swagger. What is inconvenient to work with? json encode and json decode the param string?

**Comment by slavcodev** [comment]
> What document should I rely on when changing POST to GET here

> The POST method requests that the target resource process the
   representation enclosed in the request according to the resource's
   own specific semantics. ([rfc7231](https://tools.ietf.org/html/rfc7231#section-4.3.3))

So what is target resource, and what is semantic of payload you POST for it?

**Comment by adamaltman** [comment]
You can use a body with a GET request -- elasticsearch has implemented it:

https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html

> Both HTTP GET and HTTP POST can be used to execute search with body. Since not all clients support GET with body, POST is allowed as well.

While the lib the F-E uses may support it (Axios), the browsers themselves may not:

Read this for more info:  https://github.com/axios/axios/issues/462

I believe that's why the API should be specified as GET requests, but like Elasticsearch, it should support POST too.

**Comment by slavcodev** [comment]
[About body in `GET`](https://groups.yahoo.com/neo/groups/rest-discuss/conversations/messages/9962)

> **Roy T. FieldingDec 12, 2007**
Yes. In other words, any HTTP request message is allowed to contain
a message body, and thus must parse messages with that in mind.
Server semantics for GET, however, are restricted such that a body,
if any, has no semantic meaning to the request. The requirements
on parsing are separate from the requirements on method semantics.
So, yes, you can send a body with GET, and no, it is never useful
to do so.
This is part of the layered design of HTTP/1.1 that will become
clear again once the spec is partitioned (work in progress).

**Comment by 2e3s** [comment]
> What make you decide this, contrary to your feelings?

Because `GET` is consistent.

> Are you going to read this requests or application?

The answers lies in the same area why nowadays everyone uses human readable JSON, and not a binary format. Once your code handles it, there is no difference in formats. Before that or maintaining this using binary, XML or JSON is a matter of convenience. 
It's also something which will appear in the logs by the way. Our F-E should be adjusted not to show long query strings.

> Seems it is simple to document this query param in swagger

I couldn't do that:
```
    - name: body
      in: query
      description: Dashboard request
      required: true
      schema:
        $ref: "#/definitions/DashboardRequest"
```
It doesn't render the schema. Well, that's kind of expectable, because complex query parameters with a schema is not something I would expect as a developer of Swagger. So, we have a need to modify Redoc to work with these.

**Comment by slavcodev** [comment]
> I couldn't do that

I think `body` name is reserved to `in: body`. BTW, you can about extra level, and use query param `metrics`, not body.

**Comment by 2e3s** [comment]
> The POST method requests that the target resource process the
representation enclosed in the request according to the resource's
own specific semantics.

> So what is target resource, and what is semantic of payload you POST for it?

Yes, I read this RFC before. The report is the target resource. The semantics of the request body is the report settings. The definition you've given says anything about nothing (as it seems to me) :smile:

**Comment by 2e3s** [comment]
> I think body name is reserved to in: body

No, it's the same with any name.

> BTW, you can about extra level, and use query param metrics, not body.

I'm not sure about what you suggest. I did this:
```
    - name: metrics
      in: query
      description: Dashboard request
      required: true
      schema:
        $ref: "#/definitions/DashboardRequest"
```
And get this (with no schema):
> metrics object Required

Please, clarify your suggestion. Or don't... I believe it won't work anyways.

**Comment by 2e3s** [comment]
Yes, OpenAPI 3 seems to define it as well.

File: spec/definitions/DashboardRequest.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      metricName:
+        type: object
+        description: Custom metric name which will be used in the response containing the appropriate data.
+        properties:
+          type:
+            type: string
+            description: Type of data and grid of the metric.
+            enum:
+              - approvedThroughput
+              - salesAllCount
```
Comment (adamaltman) [inline spec/definitions/DashboardRequest.yaml]: seems like `salesCount` may be better?

File: spec/definitions/DashboardResponse.yaml [hunk 4]
```diff
@@ -0,0 +1,55 @@
+type: object
+properties:
+  metrics:
+    type: object
+    properties:
+      metricName:
+        type: object
+        description: The metric name which was specified in the request with its data.
+        properties:
```
Comment (adamaltman) [inline spec/definitions/DashboardResponse.yaml]: Seems incorrect to me.  I don't see why there is a separate section to "segments", as there should only be segments.  Also, it seems there are other properties missing that should be there, like the metric name itself.

File: spec/definitions/DashboardRequest.yaml [hunk 5]
```diff
@@ -0,0 +1,27 @@
+type: object
+properties:
+  metrics:
+    type: object
+    properties:
+      metricName:
+        type: object
+        description: Custom metric name which will be used in the response containing the appropriate data.
+        properties:
```
Comment (adamaltman) [inline spec/definitions/DashboardRequest.yaml]: no date range?

File: spec/definitions/DashboardRequest.yaml (hunk 3 above)
Comment (2e3s) [reply spec/definitions/DashboardRequest.yaml]: Transaction report already has `salesCount` but it is about approved sales, and dashboard is about all sales. I wanted this to be more consistent so I didn't find anything better than this.

File: spec/definitions/DashboardResponse.yaml (hunk 4 above)
Comment (2e3s) [reply spec/definitions/DashboardResponse.yaml]: > seems there are other properties missing that should be there, like the metric name itself

The `metricName` is specified as a key for the metric settings. I didn't know how to say that "in swagger", so I explained this in description.

> I don't see why there is a separate section to "segments", as there should only be segments

I wasn't sure about the task, so I made configurable metrics (the client can choose what to show) and dependent segments.

File: spec/definitions/DashboardRequest.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+type: object
+properties:
+  segments:
+    type: object
+    description: Filtered segments for each metric
+    properties:
+      segmentName:
```
Comment (2e3s) [inline spec/definitions/DashboardRequest.yaml]:
```
{
    segments: {
        'Sales Value': {
            metric: 'salesValue',
            criteria: null
        },
        'Approval Rate': {
            metric: 'approvedThroughput',
            criteria: null
        }
    }
}
```
@adamaltman as I understood your proposition, I made everything a segment, but there is an issue (solvable with sort of prefixes): if we want a subsegment `facebook`, we need to somehow deal with the name:
```
{
    segments: {
        'Sales Value': {
            metric: 'salesValue',
            criteria: null
        },
        'Approval Rate': {
            metric: 'approvedThroughput',
            criteria: null
        }
        'facebook-salesValue': {
            metric: 'salesValue',
            criteria: {somecriteria}
        },
        'facebook-approvedThroughput': {
            metric: 'approvedThroughput',
            criteria: {somecriteria}
        }
    }
}
```

Comment (adamaltman) [reply spec/definitions/DashboardRequest.yaml]: I think this design is problematic. Let's discuss it on slack in more detail.  It violates many api design principles.

File: spec/definitions/DashboardResponse.yaml [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        type: object
+        properties:
+          value:
+            type: number
+            format: double
+            description: Segment value for the given date range.
+          previousValue:
+            type: number
+            format: double
+            description: Segment value for the previous date range (relatively to the given date range).
```
Comment (adamaltman) [inline spec/definitions/DashboardResponse.yaml]: `relative` not `relatively`

File: spec/paths/reports@dashboard.yaml [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    - name: periodEnd
+      in: query
+      description: Report end day
+      required: true
+      type: string
+      format: date-time
+    - name: metrics
+      in: query
+      description: Comma-separated list of metrics
+      required: true
```
Comment (adamaltman) [inline spec/paths/reports@dashboard.yaml]: I don't think this should be required.  If not supplied then all metrics should be provided.

File: spec/paths/reports@dashboard.yaml [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+      type: string
+      format: date-time
+    - name: segments
+      in: query
+      description: Dashboard report segments as a JSON array
+      required: true
```
Comment (adamaltman) [inline spec/paths/reports@dashboard.yaml]: Again, I don't think this should be required.  If I recall this is a complex object to create.  I think it should default to "all".

**Pull request merged**